### PR TITLE
Indirection for network config sourced addresses

### PR DESCRIPTION
**Pull request by manadart**
Adds an indirection for addresses handled by `network.ConfigSource`, and a default implementation based on stringified `net.Addr` addresses.

This moves a little more address handling logic from `api/common` to `core/network`.

Next will be to implement the `ConfigSource` and `ConfigSourceAddr` indirections with a new implementation that includes secondary address determination.

Depends on https://github.com/juju/juju/pull/12622.

## QA steps

- Bootstrap a LXD controller.
- Connect to Mongo and ensure that `db.ip.addresses.find().pretty() shows machine-sourced addresses.

## Documentation changes

None.

## Bug reference

Advances https://bugs.launchpad.net/juju/+bug/1863916

### CODE REVIEW

**Comment by manadart** [comment]
[Add indirection and default implementation for machine-sourced interface addresses](https://trello.com/c/WlULrp27/3429-add-indirection-and-default-implementation-for-machine-sourced-interface-addresses)

File: core/network/network_test.go [hunk 1]
```diff
@@ -9,11 +9,11 @@ import (
 	gc "gopkg.in/check.v1"
 
 	"github.com/juju/juju/core/network"
-	"github.com/juju/juju/testing"
+	"github.com/juju/testing"
 )
 
 type NetworkSuite struct {
-	testing.BaseSuite
+	testing.IsolationSuite
```
Comment (manadart) [inline core/network/network_test.go:16]: Just a drive-by. Using this suite is faster.

File: api/common/network_test.go [hunk 2]
```diff
@@ -430,34 +424,11 @@ func (s *NetworkSuite) TestGetObservedNetworkConfigAddressNotInCIDRFormat(c *gc.
 	s.stubConfigSource.CheckCall(c, 3, "InterfaceAddresses", "eth0")
 }
 
-func (s *NetworkSuite) TestGetObservedNetworkConfigEmptyAddressValue(c *gc.C) {
```
Comment (manadart) [inline api/common/network_test.go:433]: This logic is now pushed into the config source from the `GetObservedNetworkConfig` method. A following patch will make some changes there and re-implement this test as appropriate.

File: core/network/source.go [hunk 3]
```diff
@@ -29,7 +44,7 @@ type ConfigSource interface {
 
 	// InterfaceAddresses returns information about all addresses
 	// assigned to the network interface with the given name.
-	InterfaceAddresses(name string) ([]net.Addr, error)
+	InterfaceAddresses(name string) ([]ConfigSourceAddr, error)
```
Comment (manadart) [inline core/network/source.go:47]: I don't think this method is actually required, because when we call it, we have already acquired the actual interfaces, from which we can get the addresses directly. I will reorganise in following patches.

File: api/common/network.go [hunk 4]
```diff
@@ -158,41 +158,33 @@ func updateParentForBridgePorts(bridgeName, sysClassNetPath string, nameToConfig
...
+	interfaceName, configType string, addr corenetwork.ConfigSourceAddr,
+) (params.NetworkConfig, error) {
 	config := params.NetworkConfig{
 		ConfigType: configType,
 	}
 
-	cidrAddress := address.String()
-	if cidrAddress == "" {
-		return config, nil
+	ip := addr.IP()
```
Comment (achilleasa) [inline api/common/network.go:168]: Is there a guarantee that this function will never be called with a nil `addr` arg?

Comment (manadart) [reply api/common/network.go:168]: Based on the current implementation, yes; but I will add a guard in any case.

File: core/network/source_other.go [hunk 5]
```diff
@@ -9,25 +9,90 @@ import (
...
+
+// NewNetAddr returns a new netAddr reference representing the input IP address
+// string. No error is returned; instead the validity of input is indicated by
+// the return having a populated ip member.
+// TODO (manadart 2021-02-15): This method is exported on account of testing in
+// the api/common package where this logic used to reside and where the actual
+// detection and conversion to params is invoked.
+// The detection should also be relocated here to core/network in order that
+// the export is no longer required.
+func NewNetAddr(a string) *netAddr {
```
Comment (achilleasa) [inline core/network/source_other.go:27]: I get it that this is only used in tests but I believe it is error prone to defer the actual validation to the consumer as passing a netAddr instance around would generally imply (at least to me) that it is a valid value. Looking forward to this being removed in the future.

**Review by achilleasa — APPROVED** [review]

**Comment by manadart** [comment]
$$merge$$
